### PR TITLE
Infrastructure refresh

### LINKED ISSUE #104: Add citation information?
I am trying to cite this amazing tool in a paper revision and noticed this does not have any citation info? 
Would it be possible to mint a DOI via e.g. zenodo?
That way I could cite it properly in the final version of the paper.

### PR DESCRIPTION
**Pull request by bjlittle**
### Description

Okay... firstly, apologies :pray: 

This PR is a bit of a bloater... I had some spare cycles and decided spend them by just "going-for-it":tm: and refresh all the infrastructure of `nc-time-axis` in a oner to modernize it before the technical debt caught up with us.

I've been a bit opinionated in some places, but overall I don't think there is anything too controversial within this PR... and I have tried to theme the commits. Anyways, the effort is now sunk, so there you go; it's all about permission and forgiveness :wink:

In summary, this PR:
- drops `cirrus-ci` and migrates to GitHub Actions (GHA) :partying_face: 
  - automates citation file checking (`ci-citation`)
  - automates explicit `conda` lock file generation (`*.lock`) from top-level `*.yml` environment files (`ci-locks`) triggered on a `cron`
  - automates `MANIFEST.in` checking for the `sdist` (`ci-manifest`)
  - automates `pytest` over supported versions of `python` (`ci-tests`) with [codecov](https://about.codecov.io/) test coverage reporting
  - automates the building, testing<sup>†</sup> and publishing to PyPI of our `sdist` and binary `wheels` (`ci-wheels`) 
- adopts the workflow of leveraging `pre-commit.ci` and our hooks, rather than explicitly running the hooks ourselves in CI
- enables [dependabot](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuring-dependabot-version-updates) to automate GHA version updates, so that we easily keep pace with the community
- makes `nc-time-axis` citable (contributors please pile in and add your ORCID references)
- refreshes the `pre-commit` hooks, most notably by adopting [codespell](https://github.com/codespell-project/codespell) to automate spell checking of the codebase and documentation
- reconciles the `MANIFEST.in` with reality
- pimps the `README.md`
- adopts [PEP621](https://peps.python.org/pep-0621/) by migrating the `setup.cfg` to the `pyproject.toml`
- moves the `nc-time-axis` codebase under the `src` directory
- co-locates both `conda` YAML and `pip` core and optional dependencies under the `requirements` directory (hopefully a better hint/reminder for devs to keep them aligned)
- I also added the `CODECOV_TOKEN`, `TEST_PYPI_API_TOKEN` and `PYPI_API_TOKEN` as GHA secrets, as required by the GHAs

<sup>†</sup> - Only the binary `wheels` are tested, not the `sdist`. However, the `wheel` is built from the `sdist` using [pypa/build](https://github.com/pypa/build) and their recommended workflow.

Yeah, I think that just about covers it. Soz. :blush: 

Closes #104

### CODE REVIEW

**Comment by bjlittle** [comment]
There is something slightly odd going on here... I've never quite seen it happen before.

The GHA CI is being [triggered and executed](https://github.com/bjlittle/nc-time-axis/actions) on my forked repo development branch, but not here on the PR.

Thought anyone?

**Comment by bjlittle** [comment]
@rcomer and @spencerkclark 

Okay, just to be completely transparent here, I circumvented the branch protection on `main` to force push and undo the last commit resulting from the merge of the test PR #132.

This is pretty exceptional, however the GHA CI workflows appear to be now up and running.

I'm going to push a follow-on change to this PR branch to undo commit [f61c2c7](https://github.com/SciTools/nc-time-axis/pull/129/commits/f61c2c7dc9ada1c98a3d0c7f73ad258d9885427a), which was reactive to the GHA oddness that we were experienceing.

Hopefully, the CI wheels should spin as expected...

File: .github/workflows/ci-wheels.yml [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      - uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          user: __token__
+          password: ${{ secrets.TEST_PYPI_API_TOKEN }}
+          repository_url: https://test.pypi.org/legacy/
+          skip_existing: true
+          print_hash: true
+
+
+  publish-artifacts-pypi:
```
Comment (spencerkclark) [inline .github/workflows/ci-wheels.yml:136]: So cool that this is now automated.   We should update this line on the release instructions page in the documentation!

https://github.com/SciTools/nc-time-axis/blob/8be5d69d43af2b065d36393bb4dbe11e59fd18cb/docs/releases.rst?plain=1#L22-L23

**Review by spencerkclark — APPROVED** [review]
Well this is all pretty spectacular.  Amazing to see nc-time-axis get this kind of treatment!  All these new workflows / cleanups are great.  Many thanks for putting this together @bjlittle.  Just a couple suggestions on some places to update the documentation.  

Also please give yourself credit for this in the [release notes](https://github.com/SciTools/nc-time-axis/blob/main/docs/release-notes.rst), and if you don't mind, maybe add a note for @rcomer's work in #128 as well!

**Comment by bjlittle** [comment]
@spencerkclark Okay, over to you :+1: 

Note that, I ended up creating a `common_links.inc` files to capture common URLs, which hopefully makes the docs a wee bit easier. I also made some minor consistency changes to the docs e.g., title capitalsation, use of common links et al

Hopefully I've not broken anything, so do take a peek and ensure that's the case.

Thanks again for taking the time to review this PR. Above and beyond :beers:

File: .github/workflows/ci-wheels.yml (hunk 1 above)
Comment (bjlittle) [reply .github/workflows/ci-wheels.yml:136]: Done. See [9aca28a](https://github.com/SciTools/nc-time-axis/pull/129/commits/9aca28abc0998d3843971e3239529cf2f915063d)

File: requirements/docs.yml [hunk 2]
```diff
@@ -7,13 +7,13 @@ dependencies:
...
-  - setuptools
-  - setuptools-scm
+  - setuptools >=61
+  - setuptools-scm >=7
 
 # core dependencies
   - cftime >=1.5
   - matplotlib
-  - numpy
+  - numpy >=1.21
```
[the remark this answers is not in the document]
Comment (bjlittle) [reply requirements/docs.yml:16]: Done. See [9aca28a](https://github.com/SciTools/nc-time-axis/pull/129/commits/9aca28abc0998d3843971e3239529cf2f915063d)

**Review by spencerkclark — APPROVED** [review]
Looks good -- thanks for the additional cleanups to the documentation!

**Comment by bjlittle** [comment]
@spencerkclark Back to you 👍

**Comment by spencerkclark** [comment]
In it goes!

**Pull request merged**